lib/solvers/NetToPointPairsSolver/NetToPointPairsSolver.ts

```diff
@@ -20,14 +25,79 @@ import { seededRandom } from "lib/utils/cloneAndShuffleArray"
 export class NetToPointPairsSolver extends BaseSolver {
   unprocessedConnections: Array<SimpleRouteConnection>
   newConnections: Array<SimpleRouteConnection>
+  dsu: DSU
+  pointMap: Map<string, ConnectionPoint>
 
   constructor(
     public ogSrj: SimpleRouteJson,
     public colorMap: Record<string, string> = {},
   ) {
     super()
-    this.unprocessedConnections = [...ogSrj.connections]
     this.newConnections = []
+
+    const allPoints = ogSrj.connections.flatMap((c) => c.pointsToConnect)
+    this.pointMap = new Map<string, ConnectionPoint>()
+    for (const point of allPoints) {
+      if (point.pointId) {
+        this.pointMap.set(point.pointId, point)
+      }
+    }
+
+    const allPointIds = allPoints
+      .map((p) => p.pointId)
+      .filter((id): id is string => !!id)
+    this.dsu = new DSU(allPointIds)
+
+    const onBoardConnections: SimpleRouteConnection[] = []
+    for (const conn of ogSrj.connections) {
+      if (conn.isOffBoard) {
+        if (
+          conn.pointsToConnect.length >= 2 &&
+          conn.pointsToConnect[0].pointId &&
+          conn.pointsToConnect[1].pointId
+        ) {
+          this.dsu.union(
+            conn.pointsToConnect[0].pointId,
+            conn.pointsToConnect[1].pointId,
+          )
+        }
+      } else {
+        onBoardConnections.push(conn)
+      }
+    }
+    this.unprocessedConnections = onBoardConnections
+  }
+
+  _optimizeConnection(
+    p1: ConnectionPoint,
+    p2: ConnectionPoint,
+  ): { pointsToConnect: [ConnectionPoint, ConnectionPoint] } {
+    if (!p1.pointId || !p2.pointId) return { pointsToConnect: [p1, p2] }
+
+    const group1 = this.dsu
+      .getGroup(p1.pointId)
+      .map((id) => this.pointMap.get(id)!)
+    const group2 = this.dsu
+      .getGroup(p2.pointId)
+      .map((id) => this.pointMap.get(id)!)
+
+    let bestP1 = p1
+    let bestP2 = p2
+    let minDistance = Infinity
+
+    for (const p1Cand of group1) {
+      for (const p2Cand of group2) {
+        const dist = Math.sqrt(
+          Math.pow(p1Cand.x - p2Cand.x, 2) + Math.pow(p1Cand.y - p2Cand.y, 2),
+        )
+        if (dist < minDistance) {
+          minDistance = dist
+          bestP1 = p1Cand
+          bestP2 = p2Cand
+        }
+      }
+    }
+    return { pointsToConnect: [bestP1, bestP2] }
```

Comment:
so this general idea is fine, but what you're doing is `_findBestConnectionPointsFromDisjointSets`